File: app/src/main/java/com/sunny/youyun/wifidirect/receiver/WifiDirectBroadCastReceiver.java
package com.sunny.youyun.wifidirect.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.net.NetworkInfo;
import android.net.wifi.p2p.WifiP2pInfo;
import android.net.wifi.p2p.WifiP2pManager;

import com.sunny.youyun.wifidirect.manager.DeviceInfoManager;
import com.sunny.youyun.wifidirect.manager.WifiDirectManager;
import com.sunny.youyun.wifidirect.model.DeviceInfo;


/**
 * Wifi-Direct broadcast receiver, this receiver can catch all broadcast signal
 * send by Wifi-Direct
 * Created by qjm3662 on 2016/10/25 0025.
 */

public class WifiDirectBroadCastReceiver extends BroadcastReceiver {
    /**
     * This receiver register in manifest file , so must declare a non-param constructor
     */
    public WifiDirectBroadCastReceiver() {

    }

    @Override
    public void onReceive(Context context, Intent intent) {
        try {
            String action = intent.getAction();
            switch (action) {
                //当wifi功能打开关闭的时候会广播这个信号
                case WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION:
                    System.out.println("WIFI_P2P_STATE_CHANGED_ACTION");
                    int state = intent.getIntExtra(WifiP2pManager.EXTRA_WIFI_STATE, -1);
                    if (state == WifiP2pManager.WIFI_P2P_STATE_ENABLED) {
                        //Wifi Direct is enabled
                        WifiDirectManager.getInstance().setWifiDirectEnable(true);
                    } else {
                        //Wifi Direct is not enabled
                        WifiDirectManager.getInstance().setWifiDirectEnable(false);
                    }
                    break;

                //This action received when the peers of available device is change
                //当前可用列表发生改变时执行下面代码块，获取当前可用连接表的列表
                case WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION:
                    System.out.println("WIFI_P2P_PEERS_CHANGED_ACTION");
                    WifiDirectManager.getInstance().requestPeers(null);
                    break;

                //This action received when the connection setup or dis-setup
                //当连接建立或者断开的时候会广播该信号
                case WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION:
                    System.out.println("WIFI_P2P_CONNECTION_CHANGED_ACTION");
                    NetworkInfo networkInfo = intent.getParcelableExtra(WifiP2pManager.EXTRA_NETWORK_INFO);
                    WifiP2pInfo wifiP2pInfo = intent.getParcelableExtra(WifiP2pManager.EXTRA_WIFI_P2P_INFO);

                    //如果群组都还没建立，就直接忽略该信号
                    if (!wifiP2pInfo.groupFormed)
                        break;

                    //申请获得群组的信息，用来更新群组信息
                    WifiDirectManager.getInstance().requestGroupInfo(null);
                    //标记本设备是否是组长
                    DeviceInfoManager.getInstance().setGroupOwner(wifiP2pInfo.isGroupOwner);
                    //设置组长IP
                    DeviceInfoManager.getInstance().setGroupOwnerIp(wifiP2pInfo.groupOwnerAddress.getHostAddress());

                    WifiDirectManager.getInstance().requestConnectionInfo(null);
                    if (wifiP2pInfo.isGroupOwner) {       //群组处理，直接更新自己的设备信息
                        DeviceInfo deviceInfo = DeviceInfoManager.getInstance().getDeviceInfo();
                        //ip
                        deviceInfo.setIp(wifiP2pInfo.groupOwnerAddress.getHostAddress());
                    }
                    break;

                //当前设备的Wifi状态发生改变时会广播该信号(Wifi连接，断开)
                case WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION:
                    System.out.println("WIFI_P2P_THIS_DEVICE_CHANGED_ACTION");
                    break;
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}

File: app/src/main/java/com/sunny/youyun/wifidirect/wd_2/receiver/WifiDirectBroadCastReceiver.kt
package com.sunny.youyun.wifidirect.wd_2.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.net.NetworkInfo
import android.net.wifi.p2p.*
import android.os.Build
import com.google.gson.Gson
import com.sunny.youyun.wifidirect.wd_2.manager.WifiDirectManager
import com.sunny.youyun.wifidirect.wd_2.utils.L

/**
 * Created by sunny on 17-11-16.
 */
class WifiDirectBroadCastReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context?, intent: Intent?) {
        println("onReceive: ${intent?.action}")
        if (context == null || intent == null)
            return
        when (intent.action) {
        //this action indicate whether wifi p2p is enable
            WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION -> {
                //get the state of current device
                val state = intent.getIntExtra(WifiP2pManager.EXTRA_WIFI_STATE,
                        WifiP2pManager.WIFI_P2P_STATE_DISABLED)
                println("state: $state")
            }

        //indicate this device detail change
            WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION -> {
                val device: WifiP2pDevice = intent.getParcelableExtra(WifiP2pManager.EXTRA_WIFI_P2P_DEVICE)
                WifiDirectManager.INSTANCE
                        .updateDeviceInfo(device)
                L.json(Gson().toJson(device))
            }

        //invoke when the list of peers find, register, lost
            WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION -> {
                //api > 18 have this extra info,
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
                    val wifiP2pList: WifiP2pDeviceList
                            = intent.getParcelableExtra(WifiP2pManager.EXTRA_P2P_DEVICE_LIST)
                    WifiDirectManager.INSTANCE
                            .updateWifiP2pDeviceList(wifiP2pList)
                } else { //if the sdk version lower than 18
                    //get WifiP2pDeviceList by call WifiP2pManager.requestPeers to get
                    WifiDirectManager.INSTANCE
                            .requestPeers()
                }


            }
        //This action received when the connection setup or dis-setup
            WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION -> {
                val networkInfo = intent.getParcelableExtra<NetworkInfo>(WifiP2pManager.EXTRA_NETWORK_INFO)
                val wifiP2pInfo = intent.getParcelableExtra<WifiP2pInfo>(WifiP2pManager.EXTRA_WIFI_P2P_INFO)
                WifiDirectManager.INSTANCE
                        .updateDeviceInfo(wifiP2pInfo)
                        .updateNetWorkInfo(networkInfo)

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
                    val wifiP2pGroupInfo
                            = intent.getParcelableExtra<WifiP2pGroup>(WifiP2pManager.EXTRA_WIFI_P2P_GROUP)
                    WifiDirectManager.INSTANCE
                            .updateWifiP2pGroup(wifiP2pGroup = wifiP2pGroupInfo)
                } else {
                    WifiDirectManager.INSTANCE
                            .requestGroup()
                }
                // 如果当前还没有建立群组，则不做任何操作
                if (wifiP2pInfo.groupFormed)
                    return

            }
        }
    }

}